### PR TITLE
io: add `AsyncWriteExt::write_vectored`

### LINKED ISSUE #3676: Add vectored utility fns to AsyncWriteExt
See https://github.com/tokio-rs/tokio/issues/2692#issuecomment-813203889.

### PR DESCRIPTION
**Pull request by davidpdrsn**
Fixes #3676

cc @kaimast

### CODE REVIEW

File: tokio/src/io/util/write_vectored.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+pin_project! {
+    /// A future to write a slice of buffers to an `AsyncWrite`.
+    #[derive(Debug)]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct WriteVectored<'a, 'b, W: ?Sized> {
+        writer: &'a mut W,
+        bufs: &'a [IoSlice<'b>],
+        // Make this future `!Unpin` for compatibility with async trait methods.
+        #[pin]
+        _pin: PhantomPinned,
```
Comment (davidpdrsn) [inline tokio/src/io/util/write_vectored.rs:19]: tbh I'm not totally sure we need two lifetimes here. The test case in the docs works where both are `'a` but not sure if that could break some use cases.

Comment (Darksonn) [reply tokio/src/io/util/write_vectored.rs:19]: I prefer having both lifetimes.

**Review by Darksonn — APPROVED** [review]
This seems good to me. :+1:

Do we want a `write_all_vectored` too?

**Comment by davidpdrsn** [comment]
> Do we want a write_all_vectored too?

Yeah that would be cool. I'll make an issue to add it.

**Comment by kaimast** [comment]
Wow. That was super quick. Thanks a lot!

**Pull request merged**